src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIModelMapper.ChatCompletion.cs

```diff
@@ -0,0 +1,616 @@
+                out int value) ? value : null;
+
+            int? inputCachedTokenCount = additionalCounts.TryGetValue(
+                $"{nameof(ChatTokenUsage.InputTokenDetails)}.{nameof(ChatInputTokenUsageDetails.CachedTokenCount)}",
+                out value) ? value : null;
+
+            int? outputAudioTokenCount = additionalCounts.TryGetValue(
+                $"{nameof(ChatTokenUsage.OutputTokenDetails)}.{nameof(ChatOutputTokenUsageDetails.AudioTokenCount)}",
+                out value) ? value : null;
+
+            int? outputReasoningTokenCount = additionalCounts.TryGetValue(
+                $"{nameof(ChatTokenUsage.OutputTokenDetails)}.{nameof(ChatOutputTokenUsageDetails.ReasoningTokenCount)}",
+                out value) ? value : null;
+
+            if (inputAudioTokenCount is not null || inputCachedTokenCount is not null)
+            {
+                inputTokenUsageDetails = OpenAIChatModelFactory.ChatInputTokenUsageDetails(
+                    audioTokenCount: inputAudioTokenCount ?? 0,
+                    cachedTokenCount: inputCachedTokenCount ?? 0);
+            }
+
+            if (outputAudioTokenCount is not null || outputReasoningTokenCount is not null)
+            {
+                outputTokenUsageDetails = OpenAIChatModelFactory.ChatOutputTokenUsageDetails(
+                    audioTokenCount: outputAudioTokenCount ?? 0,
+                    reasoningTokenCount: outputReasoningTokenCount ?? 0);
+            }
+        }
+
+        return OpenAIChatModelFactory.ChatTokenUsage(
+            inputTokenCount: usageDetails.InputTokenCount ?? 0,
+            outputTokenCount: usageDetails.OutputTokenCount ?? 0,
+            totalTokenCount: usageDetails.TotalTokenCount ?? 0,
+            outputTokenDetails: outputTokenUsageDetails,
+            inputTokenDetails: inputTokenUsageDetails);
+    }
+
+    /// <summary>Converts an OpenAI role to an Extensions role.</summary>
+    private static ChatRole FromOpenAIChatRole(ChatMessageRole role) =>
+        role switch
+        {
+            ChatMessageRole.System => ChatRole.System,
+            ChatMessageRole.User => ChatRole.User,
+            ChatMessageRole.Assistant => ChatRole.Assistant,
+            ChatMessageRole.Tool => ChatRole.Tool,
+            _ => new ChatRole(role.ToString()),
+        };
+
+    /// <summary>Converts an Extensions role to an OpenAI role.</summary>
+    [return: NotNullIfNotNull("role")]
+    private static ChatMessageRole? ToOpenAIChatRole(ChatRole? role) =>
+        role switch
+        {
+            null => null,
+            _ when role == ChatRole.System => ChatMessageRole.System,
+            _ when role == ChatRole.User => ChatMessageRole.User,
+            _ when role == ChatRole.Assistant => ChatMessageRole.Assistant,
+            _ when role == ChatRole.Tool => ChatMessageRole.Tool,
+            _ => ChatMessageRole.System,
```

Comment:
System seems like the wrong default here; it's effectively promoting something unknown to the most sensitive category. It should probably be User or Assistant.

Reply:
> unless the OpenAI object model has a way to represent unrecognized role strings.

It doesn't at the moment, but it's likely to get one.